### PR TITLE
[Merged by Bors] - feat : Add Gamma distribution

### PR DESCRIPTION
**Pull request by JADekker**
Add pdf, CDF and measure of Gamma distribution.
Add proof that this is indeed a probability distribution. Add proofs that relate the various definitions.

TODO: Refactor Probability/Distributions/Exponential.lean using calls to
  the results in this file.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Probability/Distributions/Gamma.lean
Comment (loefflerd) [inline Mathlib/Probability/Distributions/Gamma.lean:253]: Use of `;` in mathlib code is discouraged. Try `refine (integral_eq_lintegral_of_nonneg_ae ?_ ?_).symm` and it doesn't create the `rfl` goal anyway

**Review by loefflerd — COMMENTED** [review]
I've left a bunch of comments. Most of these are just code-style fixes and "golf". The only comment of substance is that I think `pow_exp_integral_Ioi` should be proved in more generality (allowing complex exponents and deducing the real case from that) and moved to a different file.

**Comment by JADekker** [comment]
> I've left a bunch of comments. Most of these are just code-style fixes and "golf". The only comment of substance is that I think `pow_exp_integral_Ioi` should be proved in more generality (allowing complex exponents and deducing the real case from that) and moved to a different file.

Thank you, I'll take a good look tomorrow! Regarding some style comments: I took everything from Probability/Distributions/Exponential.lean and tried to stick as closely to that as possible; I should've probably double-checked style, I'll do that tomorrow!

File: Mathlib/Probability/Distributions/Gamma.lean
Comment (JADekker) [reply Mathlib/Probability/Distributions/Gamma.lean:86]: no, just a shameless copy-paste from Probability/Distributions/Exponential.

File: Mathlib/Probability/Distributions/Gamma.lean
Comment (JADekker) [reply Mathlib/Probability/Distributions/Gamma.lean:107]: It is a one-liner, so we can just as well remove it indeed!

File: Mathlib/Probability/Distributions/Gamma.lean
Comment (JADekker) [reply Mathlib/Probability/Distributions/Gamma.lean:254]: I just took the code from Probability/Distributions/Exponential.lean here, I didn't think too much about it here. It seems that cdf_eq_to_Real cannot infer that `gammaMeasure a r` has IsProbabilityMeasure. I got rid of the Fact's here and made the arguments explicit. To make cdf_eq_to_real happy, I have added a 'have : ...' statement to enforce this inference, but there is probably a better way.

File: Mathlib/Probability/Distributions/Gamma.lean
Comment (JADekker) [reply Mathlib/Probability/Distributions/Gamma.lean:238]: Currently I need to reference this instance explicitly in my solution to get cdf_eq_to_Real to work (see https://github.com/leanprover-community/mathlib4/pull/9408#discussion_r1440711941). What is the preferred way to fix that, so I can remove the name?

Comment (loefflerd) [reply Mathlib/Probability/Distributions/Gamma.lean:238]: It's strange that this inference doesn't happen automatically, but the workings of the type class inference system are a mystery to me. If you genuinely do need to give it a name, then there's no problem with that.

Comment (JADekker) [reply Mathlib/Probability/Distributions/Gamma.lean:238]: Yes, I am surprised as well, but I don't have a solid background in CS, so I'll leave it be for now.

File: Mathlib/Probability/Distributions/Gamma.lean
Comment (JADekker) [reply Mathlib/Probability/Distributions/Gamma.lean:121]: I'm working on this now, I'm fighting with Lean to get the type coercions working, but other than that I think it should be fine, iirc it doesn't use anything too fancy in the complexes. I'll make a commit with all changes, including this one, when it is done. Do you have a suggestion on where to put it? Close to the definition of the gamma integral?

File: Mathlib/Probability/Distributions/Gamma.lean
Comment (loefflerd) [inline Mathlib/Probability/Distributions/Gamma.lean:244]: Looks great, thanks for adding this! Can you maybe move it to `SpecialFunctions.Gamma.Basic` or somewhere like that (somewhere shortly after the definition of `Complex.Gamma`)?

I would also suggest a different name, maybe `integral_cpow_mul_exp_neg_mul_Ioi`, for this lemma; and similarly the real version should be `integral_rpow_mul_exp_neg_mul_Ioi`. (Generally we distinguish between different kinds of exponentiation depending on the type of the exponent, so `pow` means Nat exponent, and other kinds of powers are `zpow, rpow, cpow`.)

How painful would it be to deduce the real-variable version (the current `pow_exp_integral_Ioi`) from the complex version, rather than giving a separate proof as currently?

Comment (JADekker) [reply Mathlib/Probability/Distributions/Gamma.lean]: Thank you, I'll incorporate all this and make a new commit right away. Deducing should be easy, but I wanted to wait until the location was determined, so I wouldn't have to keep changing imports in Distributions/Gamma.

Comment (JADekker) [reply Mathlib/Probability/Distributions/Gamma.lean]: I'm having some issues with the coercions to give the real-variable version directly from the complex version, it is in this commit (which probably won't build because of the 'sorry')

Comment (JADekker) [reply Mathlib/Probability/Distributions/Gamma.lean]: New commit made, all points are now reflected there!

**Comment by loefflerd** [comment]
@collares, @JADekker What's going on here? I was in the middle of reviewing this branch, as specifically requested by Josha, and suddenly it's all been changed with a force-push, destroying an hour's work I just spent carefully streamlining the proofs...

**Comment by JADekker** [comment]
I'm very sorry, I was not aware that that would happen. Somehow I bogged up my git and asked for help from @collares to set back my branch; I wasn't aware of these consequences... Is there some way to fix this? Your version is the most up-to-date version, so I'm more than happy if you use that one and make a new PR with it/overwrite the current one entirely, if that is possible!

**Comment by loefflerd** [comment]
Don't worry; I manually copy-pasted the changes from my local git branch to somewhere else, downloaded the new git branch, and I'm now back on track. BTW, judging from the latest comments on zulip, it looks like I gave you misleading advice about instance parameters – sorry for that!

**Comment by JADekker** [comment]
No worries! I'm fighting with git on another branch right now, I'll make sure I don't touch this one!

**Comment by loefflerd** [comment]
I've pushed a commit which shortens the proofs in `Gamma/Basic.lean` a good deal, and also makes them quicker to compile (since there are fewer repetitions of heavyweight tactics like `simp` and `norm_cast`). I'll look at the other file next.

File: Mathlib/Probability/Distributions/Gamma.lean
Comment (loefflerd) [inline Mathlib/Probability/Distributions/Gamma.lean:186]:
```suggestion
  · exact ae_of_all _ fun b ↦ by simp only [Pi.zero_apply, gammaPdfReal_nonneg ha hr]
```

File: Mathlib/Probability/Distributions/Gamma.lean
Comment (loefflerd) [inline Mathlib/Probability/Distributions/Gamma.lean:181]: `a` and `r` can be implicitized (and `x` should probably be the last argument)

**Review by loefflerd — COMMENTED** [review]
Here are some comment. I've tried to fix the spelling / grammar errors in the docstrings for you, since they aren't your fault. (Just in case you aren't aware: you can apply all the suggestions I've made as a batch, by going to the "Files changed" view – that hopefully makes dealing with pedantic reviewers slightly less tedious :smile:)

File: Mathlib/Probability/Distributions/Gamma.lean
Comment (JADekker) [reply Mathlib/Probability/Distributions/Gamma.lean:107]: it calls that function, but it is a bit inefficient yes. The idea is that we divide by the constants that appear in the RHS from the function that we added to `SpecialFunctions.Gamma.Basic` are moved to the other side. It is an intermediate result for lintegral_gammaPdf_eq_one, I could try to refactor that proof? 
Also I agree docstring is a bit confusing here, the function doesn't really do what it promises.

File: Mathlib/Probability/Distributions/Gamma.lean
Comment (JADekker) [reply Mathlib/Probability/Distributions/Gamma.lean:170]: I think so too!

File: Mathlib/Probability/Distributions/Gamma.lean
Comment (JADekker) [reply Mathlib/Probability/Distributions/Gamma.lean:181]: Will do!

File: Mathlib/Probability/Distributions/Gamma.lean
Comment (JADekker) [reply Mathlib/Probability/Distributions/Gamma.lean:90]: Good catch! I'm familiar with 'pdf' and 'CDF', so I'll stick to that convention!

File: Mathlib/Probability/Distributions/Gamma.lean
Comment (JADekker) [reply Mathlib/Probability/Distributions/Gamma.lean]: Done, I've removed that function and refactored the proof in the most recent commit.

**Comment by JADekker** [comment]
Thank you for all the fixes, I'll try to be more critical of spelling and grammar in the future as well! I forgot about the batch apply, thank you for reminding me!

**Comment by JADekker** [comment]
@loefflerd would you like me to start refactoring Probability/Distributions/Exponential.lean and tag that along to this PR, or wait until this PR has merged? 

I think I'd also like to add the chi-squared distribution as a special case of the Gamma distribution, as it opens up the way for things like t-distributions; if there is no need for that, I'm happy to wait though!

**Comment by loefflerd** [comment]
I did some golfing on `lintegral_gammaPdf_eq_one`. This means `lintegral_eq_lintegral_Ici_add_Iio` is no longer required in this file, so I moved it back to its original home (temporarily; we can get rid of it entirely once that file gets refactored).

I'd advise waiting for a little while (until this gets merged) until you start work on the next stage. Since I am not a maintainer myself, somebody else will have to give the final OK to this, and they might ask for yet more changes; and it's always a pain to forward-port changes between multiple PR's in a series.

**Comment by loefflerd** [comment]
OK, this looks good to me now!

I suggest posting on the zulip "PR reviews" queue, asking for a maintainer to make the final merge decision.

**Comment by JADekker** [comment]
I’ll wait until it builds and then post it there! Thank you! Feel free to add yourself to the copyright statement, if you like!

File: Mathlib/Probability/Distributions/Gamma.lean
Comment (RemyDegenne) [inline Mathlib/Probability/Distributions/Gamma.lean:24]:
```suggestion
* `gammaCdfReal`: the CDF given by the definition of CDF in `ProbabilityTheory.Cdf` applied to the
  Gamma measure.
```

File: Mathlib/Probability/Distributions/Gamma.lean
Comment (RemyDegenne) [inline Mathlib/Probability/Distributions/Gamma.lean:43]:
```suggestion
/-- The pdf of the gamma distribution depending on its scale and rate-/
```

File: Mathlib/Probability/Distributions/Gamma.lean
Comment (RemyDegenne) [inline Mathlib/Probability/Distributions/Gamma.lean:133]:
```suggestion
lemma gammaCdfReal_eq_integral {a r : ℝ} (ha : 0 < a) (hr : 0 < r) (x : ℝ) :
```

**Review by RemyDegenne — APPROVED** [review]
LGTM, thanks!
Here are 3 very minor comments. You can merge yourself after resolving them.

bors d+

**Pull request closed without merging**
